examples/seismic/tutorials/06_skimage_tv.ipynb

```diff
@@ -0,0 +1,416 @@
+    "\n",
+    "    # Generate synthetic receiver data from true model\n",
+    "    true_d, _, _ = solver.forward(src=src, m=true_model.m)\n",
+    "\n",
+    "    return src, true_d, nt, solver\n",
+    "\n",
+    "# Define a type to store the functional and gradient.\n",
+    "class fg_pair:\n",
+    "    def __init__(self, f, g):\n",
+    "        self.f = f\n",
+    "        self.g = g\n",
+    "    \n",
+    "    def __add__(self, other):\n",
+    "        f = self.f + other.f\n",
+    "        g = self.g + other.g\n",
+    "        \n",
+    "        return fg_pair(f, g)\n",
+    "    \n",
+    "    def __radd__(self, other):\n",
+    "        if other == 0:\n",
+    "            return self\n",
+    "        else:\n",
+    "            return self.__add__(other)\n",
+    "\n",
+    "from devito import Function, clear_cache\n",
+    "\n",
+    "# Create FWI gradient kernel for a single shot\n",
+    "def fwi_gradient_i(param):\n",
+    "    # Need to clear the workers cache.\n",
+    "    clear_cache()\n",
+    "\n",
+    "    # Load the current model and the shot data for this worker.\n",
+    "    # Note, unlike the serial example the model is not passed in\n",
+    "    # as an argument. Broadcasting large datasets is considered\n",
+    "    # a programming anti-pattern and at the time of writing it\n",
+    "    # it only worked relaiably with Dask master. Therefore, the\n",
+    "    # the model is communicated via a file.\n",
+    "    model0 = get_current_model(param)\n",
+    "    src, rec, nt, solver = get_data(param)\n",
+    "    \n",
+    "    # Create symbols to hold the gradient and the mismatch between\n",
+    "    # the 'measured' and simulated data.\n",
+    "    grad = Function(name=\"grad\", grid=model0.grid)\n",
+    "    residual = Receiver(name='rec', grid=model0.grid, ntime=nt, coordinates=rec.coordinates.data)\n",
+    "    \n",
+    "    # Compute simulated data and full forward wavefield u0\n",
+    "    d, u0, _ = solver.forward(src=src, m=model0.m, save=True)\n",
+    "        \n",
+    "    # Compute the data mismatch (residual) and objective function  \n",
+    "    residual.data[:] = d.data[:] - rec.data[:]\n",
+    "    f = .5*np.linalg.norm(residual.data.flatten())**2\n",
+    "    \n",
+    "    # Compute gradient using the adjoint-state method. Note, this\n",
+    "    # backpropogates the data mismatch through the model.\n",
+    "    solver.gradient(rec=residual, u=u0, m=model0.m, grad=grad)\n",
+    "    \n",
+    "    # Taking here to avoid a (probably overzealous) destructor deleting\n",
+    "    # the gradient before Dask has had a chance to communicate it.\n",
+    "    g = np.array(grad.data[:])\n",
+    "    \n",
+    "    # return the objective functional and gradient.\n",
+    "    return fg_pair(f, g)\n",
+    "\n",
+    "import numpy as np\n",
+    "from distributed import Client\n",
+    "\n",
+    "# Dumps the model to disk; workers will pick this up when they need it.\n",
```

Comment:
utility methods can go into a tools module

Reply:
Agreed; will be part of a future PR.